File: clients/swift/Tests/KgdClientTests/ClientTests.swift
/// Tests for kgdclient types and protocol encoding.

import XCTest
import MessagePack
@testable import KgdClient

final class ClientTests: XCTestCase {

    // MARK: - Anchor tests

    func testAnchorAbsolute() {
        let a = Anchor(type: "absolute", row: 5, col: 10)
        let d = a.toDictionary()
        XCTAssertEqual(d["type"] as? String, "absolute")
        XCTAssertEqual(d["row"] as? Int, 5)
        XCTAssertEqual(d["col"] as? Int, 10)
        XCTAssertNil(d["pane_id"])
        XCTAssertNil(d["win_id"])
        XCTAssertNil(d["buf_line"])
    }

    func testAnchorPane() {
        let a = Anchor(type: "pane", paneID: "%0", row: 2, col: 3)
        let d = a.toDictionary()
        XCTAssertEqual(d["type"] as? String, "pane")
        XCTAssertEqual(d["pane_id"] as? String, "%0")
        XCTAssertEqual(d["row"] as? Int, 2)
        XCTAssertEqual(d["col"] as? Int, 3)
    }

    func testAnchorNvimWin() {
        let a = Anchor(type: "nvim_win", winID: 1000, bufLine: 5, col: 0)
        let d = a.toDictionary()
        XCTAssertEqual(d["type"] as? String, "nvim_win")
        XCTAssertEqual(d["win_id"] as? Int, 1000)
        XCTAssertEqual(d["buf_line"] as? Int, 5)
        // col=0 should be omitted
        XCTAssertNil(d["col"])
    }

    func testAnchorOmitsZeroFields() {
        let a = Anchor(type: "absolute")
        let d = a.toDictionary()
        XCTAssertEqual(d.count, 1)
        XCTAssertEqual(d["type"] as? String, "absolute")
    }

    func testAnchorMsgpackRoundtrip() {
        let a = Anchor(type: "pane", paneID: "%1", row: 10, col: 20)
        let packed = a.toMsgpack()
        guard let dict = packed.dictionaryValue else {
            XCTFail("expected map")
            return
        }
        XCTAssertEqual(dict[.string("type")]?.stringValue, "pane")
        XCTAssertEqual(dict[.string("pane_id")]?.stringValue, "%1")
        XCTAssertEqual(dict[.string("row")]?.int64Value, 10)
        XCTAssertEqual(dict[.string("col")]?.int64Value, 20)
        // Zero fields should not be present
        XCTAssertNil(dict[.string("win_id")])
        XCTAssertNil(dict[.string("buf_line")])
    }

    // MARK: - Color tests

    func testColorDefaults() {
        let c = Color()
        XCTAssertEqual(c.r, 0)
        XCTAssertEqual(c.g, 0)
        XCTAssertEqual(c.b, 0)
    }

    func testColorValues() {
        let c = Color(r: 65535, g: 32768, b: 0)
        XCTAssertEqual(c.r, 65535)
        XCTAssertEqual(c.g, 32768)
        XCTAssertEqual(c.b, 0)
    }

    func testColorFromMsgpack() {
        let value: MessagePackValue = .map([
            .string("r"): .int(255),
            .string("g"): .int(128),
            .string("b"): .int(64),
        ])
        let c = Color.from(msgpack: value)
        XCTAssertEqual(c.r, 255)
        XCTAssertEqual(c.g, 128)
        XCTAssertEqual(c.b, 64)
    }

    func testColorFromMsgpackMissingFields() {
        let value: MessagePackValue = .map([:])
        let c = Color.from(msgpack: value)
        XCTAssertEqual(c.r, 0)
        XCTAssertEqual(c.g, 0)
        XCTAssertEqual(c.b, 0)
    }

    // MARK: - PlacementInfo tests

    func testPlacementInfoFromMsgpack() {
        let value: MessagePackValue = .map([
            .string("placement_id"): .int(42),
            .string("client_id"): .string("abc-123"),
            .string("handle"): .int(7),
            .string("visible"): .bool(true),
            .string("row"): .int(5),
            .string("col"): .int(10),
        ])
        let p = PlacementInfo.from(msgpack: value)
        XCTAssertNotNil(p)
        XCTAssertEqual(p?.placementID, 42)
        XCTAssertEqual(p?.clientID, "abc-123")
        XCTAssertEqual(p?.handle, 7)
        XCTAssertEqual(p?.visible, true)
        XCTAssertEqual(p?.row, 5)
        XCTAssertEqual(p?.col, 10)
    }

    func testPlacementInfoFromInvalidMsgpack() {
        let value: MessagePackValue = .string("not a map")
        XCTAssertNil(PlacementInfo.from(msgpack: value))
    }

    // MARK: - StatusResult tests

    func testStatusResultFromMsgpack() {
        let value: MessagePackValue = .map([
            .string("clients"): .int(3),
            .string("placements"): .int(10),
            .string("images"): .int(5),
            .string("cols"): .int(80),
            .string("rows"): .int(24),
        ])
        let s = StatusResult.from(msgpack: value)
        XCTAssertEqual(s.clients, 3)
        XCTAssertEqual(s.placements, 10)
        XCTAssertEqual(s.images, 5)
        XCTAssertEqual(s.cols, 80)
        XCTAssertEqual(s.rows, 24)
    }

    func testStatusResultDefaults() {
        let s = StatusResult()
        XCTAssertEqual(s.clients, 0)
        XCTAssertEqual(s.placements, 0)
        XCTAssertEqual(s.images, 0)
        XCTAssertEqual(s.cols, 0)
        XCTAssertEqual(s.rows, 0)
    }

    // MARK: - HelloResult tests

    func testHelloResultFromMsgpack() {
        let value: MessagePackValue = .map([
            .string("client_id"): .string("uuid-here"),
            .string("cols"): .int(120),
            .string("rows"): .int(40),
            .string("cell_width"): .int(8),
            .string("cell_height"): .int(16),
            .string("in_tmux"): .bool(true),
            .string("fg"): .map([
                .string("r"): .int(65535),
                .string("g"): .int(65535),
                .string("b"): .int(65535),
            ]),
            .string("bg"): .map([
                .string("r"): .int(0),
                .string("g"): .int(0),
                .string("b"): .int(0),
            ]),
        ])
        let h = HelloResult.from(msgpack: value)
        XCTAssertEqual(h.clientID, "uuid-here")
        XCTAssertEqual(h.cols, 120)
        XCTAssertEqual(h.rows, 40)
        XCTAssertEqual(h.cellWidth, 8)
        XCTAssertEqual(h.cellHeight, 16)
        XCTAssertTrue(h.inTmux)
        XCTAssertEqual(h.fg, Color(r: 65535, g: 65535, b: 65535))
        XCTAssertEqual(h.bg, Color(r: 0, g: 0, b: 0))
    }

    // MARK: - Protocol encoding tests

    func testEncodeRequestWithParams() {
        let params: MessagePackValue = .map([
            .string("client_type"): .string("test"),
            .string("pid"): .int(1234),
        ])
        let data = encodeRequest(msgID: 1, method: "hello", params: params)
        // Decode and verify structure
        let value = try! unpackFirst(data)
        guard let arr = value.arrayValue else {
            XCTFail("expected array")
            return
        }
        XCTAssertEqual(arr.count, 4)
        XCTAssertEqual(arr[0].uint64Value, 0) // request type
        XCTAssertEqual(arr[1].uint64Value, 1) // msgid
        XCTAssertEqual(arr[2].stringValue, "hello") // method
        // params should be an array with one element (the map)
        guard let paramsArr = arr[3].arrayValue else {
            XCTFail("expected params array")
            return
        }
        XCTAssertEqual(paramsArr.count, 1)
        XCTAssertNotNil(paramsArr[0].dictionaryValue)
    }

    func testEncodeRequestWithoutParams() {
        let data = encodeRequest(msgID: 5, method: "status", params: nil)
        let value = try! unpackFirst(data)
        guard let arr = value.arrayValue else {
            XCTFail("expected array")
            return
        }
        XCTAssertEqual(arr.count, 4)
        XCTAssertEqual(arr[0].uint64Value, 0)
        XCTAssertEqual(arr[1].uint64Value, 5)
        XCTAssertEqual(arr[2].stringValue, "status")
        guard let paramsArr = arr[3].arrayValue else {
            XCTFail("expected params array")
            return
        }
        XCTAssertEqual(paramsArr.count, 0) // empty params
    }

    func testEncodeNotificationWithParams() {
        let params: MessagePackValue = .map([
            .string("win_id"): .int(42),
        ])
        let data = encodeNotification(method: "unregister_win", params: params)
        let value = try! unpackFirst(data)
        guard let arr = value.arrayValue else {
            XCTFail("expected array")
            return
        }
        XCTAssertEqual(arr.count, 3)
        XCTAssertEqual(arr[0].uint64Value, 2) // notification type
        XCTAssertEqual(arr[1].stringValue, "unregister_win")
        guard let paramsArr = arr[2].arrayValue else {
            XCTFail("expected params array")
            return
        }
        XCTAssertEqual(paramsArr.count, 1)
    }

    func testEncodeNotificationWithoutParams() {
        let data = encodeNotification(method: "stop", params: nil)
        let value = try! unpackFirst(data)
        guard let arr = value.arrayValue else {
            XCTFail("expected array")
            return
        }
        XCTAssertEqual(arr.count, 3)
        XCTAssertEqual(arr[0].uint64Value, 2)
        XCTAssertEqual(arr[1].stringValue, "stop")
        guard let paramsArr = arr[2].arrayValue else {
            XCTFail("expected params array")
            return
        }
        XCTAssertEqual(paramsArr.count, 0)
    }

    // MARK: - Protocol parsing tests

    func testParseResponseOk() throws {
        let msg: MessagePackValue = .array([
            .int(1), // response
            .uint(3), // msgid
            .nil, // no error
            .map([.string("handle"): .int(42)]), // result
        ])
        let parsed = try parseMessage(msg)
        guard case .response(let resp) = parsed else {
            XCTFail("expected response")
            return
        }
        XCTAssertEqual(resp.msgID, 3)
        XCTAssertTrue(resp.error.isNil)
        XCTAssertEqual(resp.result.dictionaryValue?[.string("handle")]?.int64Value, 42)
    }

    func testParseResponseError() throws {
        let msg: MessagePackValue = .array([
            .int(1),
            .uint(7),
            .map([.string("message"): .string("not found")]),
            .nil,
        ])
        let parsed = try parseMessage(msg)
        guard case .response(let resp) = parsed else {
            XCTFail("expected response")
            return
        }
        XCTAssertEqual(resp.msgID, 7)
        XCTAssertFalse(resp.error.isNil)
        XCTAssertEqual(resp.error.dictionaryValue?[.string("message")]?.stringValue, "not found")
    }

    func testParseNotification() throws {
        let msg: MessagePackValue = .array([
            .int(2), // notification
            .string("evicted"),
            .array([.map([.string("handle"): .int(99)])]),
        ])
        let parsed = try parseMessage(msg)
        guard case .notification(let notif) = parsed else {
            XCTFail("expected notification")
            return
        }
        XCTAssertEqual(notif.method, "evicted")
        XCTAssertEqual(notif.params.dictionaryValue?[.string("handle")]?.int64Value, 99)
    }

    func testParseInvalidMessage() {
        let msg: MessagePackValue = .string("not an array")
        XCTAssertThrowsError(try parseMessage(msg))
    }

    func testParseUnknownType() {
        let msg: MessagePackValue = .array([.int(9), .string("x"), .array([])])
        XCTAssertThrowsError(try parseMessage(msg))
    }

    // MARK: - StreamDecoder tests

    func testStreamDecoderSingleMessage() {
        let decoder = StreamDecoder()
        let value: MessagePackValue = .array([.int(1), .uint(0), .nil, .string("ok")])
        let data = pack(value)
        decoder.append(data)
        let decoded = decoder.nextValue()
        XCTAssertNotNil(decoded)
        XCTAssertEqual(decoded?.arrayValue?.count, 4)
        // No more values
        XCTAssertNil(decoder.nextValue())
    }

    func testStreamDecoderMultipleMessages() {
        let decoder = StreamDecoder()
        let v1: MessagePackValue = .array([.int(1), .uint(0), .nil, .int(1)])
        let v2: MessagePackValue = .array([.int(1), .uint(1), .nil, .int(2)])
        var data = pack(v1)
        data.append(pack(v2))
        decoder.append(data)

        let d1 = decoder.nextValue()
        XCTAssertNotNil(d1)
        XCTAssertEqual(d1?.arrayValue?[1].uint64Value, 0)

        let d2 = decoder.nextValue()
        XCTAssertNotNil(d2)
        XCTAssertEqual(d2?.arrayValue?[1].uint64Value, 1)

        XCTAssertNil(decoder.nextValue())
    }

    func testStreamDecoderPartialData() {
        let decoder = StreamDecoder()
        let value: MessagePackValue = .array([.int(1), .uint(0), .nil, .string("hello")])
        let data = pack(value)

        // Feed only half the data
        let half = data.count / 2
        decoder.append(data.prefix(half))
        XCTAssertNil(decoder.nextValue()) // not enough data

        // Feed the rest
        decoder.append(data.suffix(from: half))
        let decoded = decoder.nextValue()
        XCTAssertNotNil(decoded)
    }

    // MARK: - Socket path resolution tests

    func testResolveSocketPathOverride() {
        let path = resolveSocketPath("/custom/path.sock")
        XCTAssertEqual(path, "/custom/path.sock")
    }

    // MARK: - Options tests

    func testOptionsDefaults() {
        let opts = Options()
        XCTAssertEqual(opts.socketPath, "")
        XCTAssertEqual(opts.sessionID, "")
        XCTAssertEqual(opts.clientType, "")
        XCTAssertEqual(opts.label, "")
        XCTAssertTrue(opts.autoLaunch)
    }

    func testOptionsCustom() {
        let opts = Options(
            socketPath: "/tmp/test.sock",
            sessionID: "sess-1",
            clientType: "test",
            label: "my-label",
            autoLaunch: false
        )
        XCTAssertEqual(opts.socketPath, "/tmp/test.sock")
        XCTAssertEqual(opts.sessionID, "sess-1")
        XCTAssertEqual(opts.clientType, "test")
        XCTAssertEqual(opts.label, "my-label")
        XCTAssertFalse(opts.autoLaunch)
    }

    // MARK: - PlaceOptions tests

    func testPlaceOptionsDefaults() {
        let opts = PlaceOptions()
        XCTAssertEqual(opts.srcX, 0)
        XCTAssertEqual(opts.srcY, 0)
        XCTAssertEqual(opts.srcW, 0)
        XCTAssertEqual(opts.srcH, 0)
        XCTAssertEqual(opts.zIndex, 0)
    }
}

File: clients/dotnet/KgdClient.Tests/ClientTests.cs
using Kgd;

namespace Kgd.Tests;

public class AnchorTests
{
    [Fact]
    public void Absolute_OnlyIncludesType()
    {
        var anchor = new Anchor(Type: "absolute");
        var dict = anchor.ToDict();

        Assert.Single(dict);
        Assert.Equal("absolute", dict["type"]);
    }

    [Fact]
    public void Absolute_WithRowCol()
    {
        var anchor = new Anchor(Type: "absolute", Row: 5, Col: 10);
        var dict = anchor.ToDict();

        Assert.Equal("absolute", dict["type"]);
        Assert.Equal(5, dict["row"]);
        Assert.Equal(10, dict["col"]);
        Assert.Equal(3, dict.Count);
    }

    [Fact]
    public void Pane_WithPaneIdRowCol()
    {
        var anchor = new Anchor(Type: "pane", PaneId: "%0", Row: 2, Col: 3);
        var dict = anchor.ToDict();

        Assert.Equal("pane", dict["type"]);
        Assert.Equal("%0", dict["pane_id"]);
        Assert.Equal(2, dict["row"]);
        Assert.Equal(3, dict["col"]);
        Assert.Equal(4, dict.Count);
    }

    [Fact]
    public void NvimWin_WithWinIdBufLine()
    {
        var anchor = new Anchor(Type: "nvim_win", WinId: 1000, BufLine: 5, Col: 0);
        var dict = anchor.ToDict();

        Assert.Equal("nvim_win", dict["type"]);
        Assert.Equal(1000, dict["win_id"]);
        Assert.Equal(5, dict["buf_line"]);
        // Col=0 should be omitted
        Assert.False(dict.ContainsKey("col"));
        Assert.Equal(3, dict.Count);
    }

    [Fact]
    public void OmitsAllZeroFields()
    {
        var anchor = new Anchor(Type: "absolute");
        var dict = anchor.ToDict();

        Assert.False(dict.ContainsKey("pane_id"));
        Assert.False(dict.ContainsKey("win_id"));
        Assert.False(dict.ContainsKey("buf_line"));
        Assert.False(dict.ContainsKey("row"));
        Assert.False(dict.ContainsKey("col"));
    }
}

public class ColorTests
{
    [Fact]
    public void Defaults_AreZero()
    {
        var color = new Color();
        Assert.Equal(0, color.R);
        Assert.Equal(0, color.G);
        Assert.Equal(0, color.B);
    }

    [Fact]
    public void Values_ArePreserved()
    {
        var color = new Color(R: 65535, G: 32768, B: 0);
        Assert.Equal(65535, color.R);
        Assert.Equal(32768, color.G);
        Assert.Equal(0, color.B);
    }
}

public class PlacementInfoTests
{
    [Fact]
    public void Defaults()
    {
        var info = new PlacementInfo();
        Assert.Equal(0, info.PlacementId);
        Assert.Equal("", info.ClientId);
        Assert.Equal(0, info.Handle);
        Assert.False(info.Visible);
        Assert.Equal(0, info.Row);
        Assert.Equal(0, info.Col);
    }

    [Fact]
    public void WithValues()
    {
        var info = new PlacementInfo(PlacementId: 42, ClientId: "abc", Handle: 7, Visible: true, Row: 3, Col: 5);
        Assert.Equal(42, info.PlacementId);
        Assert.Equal("abc", info.ClientId);
        Assert.Equal(7, info.Handle);
        Assert.True(info.Visible);
        Assert.Equal(3, info.Row);
        Assert.Equal(5, info.Col);
    }
}

public class StatusResultTests
{
    [Fact]
    public void Defaults()
    {
        var status = new StatusResult();
        Assert.Equal(0, status.Clients);
        Assert.Equal(0, status.Placements);
        Assert.Equal(0, status.Images);
        Assert.Equal(0, status.Cols);
        Assert.Equal(0, status.Rows);
    }
}

public class OptionsTests
{
    [Fact]
    public void Defaults()
    {
        var opts = new Options();
        Assert.Equal("", opts.SocketPath);
        Assert.Equal("", opts.SessionId);
        Assert.Equal("", opts.ClientType);
        Assert.Equal("", opts.Label);
        Assert.True(opts.AutoLaunch);
    }
}

public class PlaceOptsTests
{
    [Fact]
    public void Defaults()
    {
        var opts = new PlaceOpts();
        Assert.Equal(0, opts.SrcX);
        Assert.Equal(0, opts.SrcY);
        Assert.Equal(0, opts.SrcW);
        Assert.Equal(0, opts.SrcH);
        Assert.Equal(0, opts.ZIndex);
    }
}

public class ProtocolTests
{
    [Fact]
    public void EncodeRequest_WithParams_RoundTrips()
    {
        var data = Protocol.EncodeRequest(1, "hello", new Dictionary<string, object>
        {
            ["client_type"] = "test",
            ["pid"] = 1234,
        });

        var consumed = Protocol.TryDecode(data, out var message);
        Assert.True(consumed > 0);
        Assert.NotNull(message);

        var arr = Assert.IsType<object?[]>(message);
        Assert.Equal(4, arr.Length);
        Assert.Equal(0, Protocol.ConvertToInt(arr[0])); // Request
        Assert.Equal(1, Protocol.ConvertToInt(arr[1])); // msgId
        Assert.Equal("hello", arr[2]?.ToString());

        // params array
        var paramsArr = Assert.IsType<object?[]>(arr[3]);
        Assert.Single(paramsArr);
    }

    [Fact]
    public void EncodeRequest_NullParams_HasEmptyArray()
    {
        var data = Protocol.EncodeRequest(0, "status", null);

        var consumed = Protocol.TryDecode(data, out var message);
        Assert.True(consumed > 0);

        var arr = Assert.IsType<object?[]>(message);
        var paramsArr = Assert.IsType<object?[]>(arr[3]);
        Assert.Empty(paramsArr);
    }

    [Fact]
    public void EncodeNotification_WithParams()
    {
        var data = Protocol.EncodeNotification("stop", null);

        var consumed = Protocol.TryDecode(data, out var message);
        Assert.True(consumed > 0);

        var arr = Assert.IsType<object?[]>(message);
        Assert.Equal(3, arr.Length);
        Assert.Equal(2, Protocol.ConvertToInt(arr[0])); // Notification
        Assert.Equal("stop", arr[1]?.ToString());
    }

    [Fact]
    public void TryDecode_IncompleteData_ReturnsZero()
    {
        var data = new byte[] { 0x94 }; // start of a 4-element array but no content
        var consumed = Protocol.TryDecode(data, out var message);
        Assert.Equal(0, consumed);
        Assert.Null(message);
    }

    [Fact]
    public void TryDecode_EmptyBuffer_ReturnsZero()
    {
        var consumed = Protocol.TryDecode(ReadOnlyMemory<byte>.Empty, out var message);
        Assert.Equal(0, consumed);
        Assert.Null(message);
    }

    [Fact]
    public void ParseResponse_ValidResponse()
    {
        // Simulate a response: [1, 42, nil, {"handle": 7}]
        var data = Protocol.EncodeRequest(0, "test", null); // just to have something encodable

        // Build a mock response by hand using MessagePack
        var writer = new ArrayBufferWriter();
        var msgpackWriter = new MessagePack.MessagePackWriter(writer);
        msgpackWriter.WriteArrayHeader(4);
        msgpackWriter.Write(1); // response type
        msgpackWriter.Write(42); // msgId
        msgpackWriter.WriteNil(); // no error
        msgpackWriter.WriteMapHeader(1);
        msgpackWriter.Write("handle");
        msgpackWriter.Write(7);
        msgpackWriter.Flush();

        var consumed = Protocol.TryDecode(writer.WrittenSpan.ToArray(), out var message);
        Assert.True(consumed > 0);

        var response = Protocol.ParseResponse(message);
        Assert.NotNull(response);
        Assert.Equal(42, response.Value.MsgId);
        Assert.Null(response.Value.Error);

        var result = Protocol.ToDictionary(response.Value.Result);
        Assert.NotNull(result);
        Assert.Equal(7, Protocol.ConvertToInt(result["handle"]));
    }

    [Fact]
    public void ParseNotification_ValidNotification()
    {
        // Build a notification: [2, "evicted", [{handle: 5}]]
        var writer = new ArrayBufferWriter();
        var msgpackWriter = new MessagePack.MessagePackWriter(writer);
        msgpackWriter.WriteArrayHeader(3);
        msgpackWriter.Write(2); // notification type
        msgpackWriter.Write("evicted");
        msgpackWriter.WriteArrayHeader(1);
        msgpackWriter.WriteMapHeader(1);
        msgpackWriter.Write("handle");
        msgpackWriter.Write(5);
        msgpackWriter.Flush();

        var consumed = Protocol.TryDecode(writer.WrittenSpan.ToArray(), out var message);
        Assert.True(consumed > 0);

        var notification = Protocol.ParseNotification(message);
        Assert.NotNull(notification);
        Assert.Equal("evicted", notification.Value.Method);
        Assert.NotNull(notification.Value.Params);
        Assert.Equal(5, Protocol.ConvertToInt(notification.Value.Params["handle"]));
    }

    [Fact]
    public void ConvertToInt_HandlesVariousTypes()
    {
        Assert.Equal(42, Protocol.ConvertToInt((byte)42));
        Assert.Equal(42, Protocol.ConvertToInt((sbyte)42));
        Assert.Equal(42, Protocol.ConvertToInt((short)42));
        Assert.Equal(42, Protocol.ConvertToInt((ushort)42));
        Assert.Equal(42, Protocol.ConvertToInt(42));
        Assert.Equal(42, Protocol.ConvertToInt((uint)42));
        Assert.Equal(42, Protocol.ConvertToInt((long)42));
        Assert.Equal(42, Protocol.ConvertToInt((ulong)42));
        Assert.Equal(0, Protocol.ConvertToInt(null));
        Assert.Equal(0, Protocol.ConvertToInt("not a number"));
    }

    [Fact]
    public void ConvertToBool_HandlesTypes()
    {
        Assert.True(Protocol.ConvertToBool(true));
        Assert.False(Protocol.ConvertToBool(false));
        Assert.False(Protocol.ConvertToBool(null));
        Assert.False(Protocol.ConvertToBool(42));
    }

    [Fact]
    public void ParseColor_ValidDict()
    {
        var dict = new Dictionary<object, object>
        {
            { "r", 65535 },
            { "g", 32768 },
            { "b", 0 },
        };

        var color = Protocol.ParseColor(dict);
        Assert.Equal(65535, color.R);
        Assert.Equal(32768, color.G);
        Assert.Equal(0, color.B);
    }

    [Fact]
    public void ParseColor_Null_ReturnsDefault()
    {
        var color = Protocol.ParseColor(null);
        Assert.Equal(0, color.R);
        Assert.Equal(0, color.G);
        Assert.Equal(0, color.B);
    }
}

public class SocketPathTests
{
    [Fact]
    public void ConfiguredPath_TakesPrecedence()
    {
        var path = KgdClient.ResolveSocketPath("/custom/path.sock");
        Assert.Equal("/custom/path.sock", path);
    }

    [Fact]
    public void EmptyConfiguredPath_FallsToEnvOrDefault()
    {
        // Just verify it returns a non-empty string (actual path depends on env)
        var path = KgdClient.ResolveSocketPath("");
        Assert.False(string.IsNullOrEmpty(path));
    }
}

public class MultipleMessagesTests
{
    [Fact]
    public void TryDecode_MultipleMessages_InOneBuffer()
    {
        var msg1 = Protocol.EncodeNotification("evicted", new Dictionary<string, object> { ["handle"] = 1 });
        var msg2 = Protocol.EncodeNotification("evicted", new Dictionary<string, object> { ["handle"] = 2 });

        // Concatenate
        var combined = new byte[msg1.Length + msg2.Length];
        Buffer.BlockCopy(msg1, 0, combined, 0, msg1.Length);
        Buffer.BlockCopy(msg2, 0, combined, msg1.Length, msg2.Length);

        // Decode first message
        var consumed1 = Protocol.TryDecode(combined, out var message1);
        Assert.True(consumed1 > 0);
        Assert.NotNull(message1);

        // Decode second message from remaining bytes
        var consumed2 = Protocol.TryDecode(combined.AsMemory(consumed1), out var message2);
        Assert.True(consumed2 > 0);
        Assert.NotNull(message2);

        // Verify both are notifications
        var n1 = Protocol.ParseNotification(message1);
        var n2 = Protocol.ParseNotification(message2);
        Assert.NotNull(n1);
        Assert.NotNull(n2);
        Assert.Equal(1, Protocol.ConvertToInt(n1.Value.Params!["handle"]));
        Assert.Equal(2, Protocol.ConvertToInt(n2.Value.Params!["handle"]));
    }
}

public class RequestEncodingTests
{
    [Fact]
    public void EncodeRequest_WithBinaryData()
    {
        var imageData = new byte[] { 0x89, 0x50, 0x4E, 0x47 }; // PNG magic bytes
        var data = Protocol.EncodeRequest(5, "upload", new Dictionary<string, object>
        {
            ["data"] = imageData,
            ["format"] = "png",
            ["width"] = 100,
            ["height"] = 200,
        });

        var consumed = Protocol.TryDecode(data, out var message);
        Assert.True(consumed > 0);

        var arr = Assert.IsType<object?[]>(message);
        Assert.Equal(5, Protocol.ConvertToInt(arr[1])); // msgId
        Assert.Equal("upload", arr[2]?.ToString());
    }

    [Fact]
    public void EncodeNotification_WithParams_RoundTrips()
    {
        var data = Protocol.EncodeNotification("register_win", new Dictionary<string, object>
        {
            ["win_id"] = 42,
            ["pane_id"] = "%0",
            ["top"] = 1,
            ["left"] = 2,
            ["width"] = 80,
            ["height"] = 24,
            ["scroll_top"] = 0,
        });

        var consumed = Protocol.TryDecode(data, out var message);
        Assert.True(consumed > 0);

        var notification = Protocol.ParseNotification(message);
        Assert.NotNull(notification);
        Assert.Equal("register_win", notification.Value.Method);
        Assert.NotNull(notification.Value.Params);
        Assert.Equal(42, Protocol.ConvertToInt(notification.Value.Params["win_id"]));
        Assert.Equal("%0", notification.Value.Params["pane_id"]?.ToString());
    }
}
